### PR TITLE
Rename party classes

### PR DESCRIPTION
**Pull request by rnicoll**
* Rename AnonymousPartyAndPath to VerifiedAnonymousParty
* Rename PartyAndCertificate to VerifiedParty

### CODE REVIEW

**Comment by rnicoll** [comment]
Trying to get the API refactor into M14 if I can, to minimize change in M15.

File: core/src/main/kotlin/net/corda/core/identity/VerifiedParty.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        else
+            false
+    }
+
+    override fun hashCode(): Int = party.hashCode()
+    /**
+     * Convert this party and certificate into an anomymised identity. This exists primarily for example cases which
+     * want to use well known identities as if they're anonymous identities.
+     */
+    fun toAnonymisedIdentity(): VerifiedAnonymousParty {
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/identity/VerifiedParty.kt]: `anonymise`?

File: core/src/main/kotlin/net/corda/core/identity/VerifiedParty.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        require(subjectX500Name == certificate.subject) { "Certificate path must end with the subject ${certificate.subject}" }
+    }
+
+    val name: X500Name
+        get() = party.name
+    val owningKey: PublicKey
+        get() = party.owningKey
+
+    override fun equals(other: Any?): Boolean {
+        return if (other is VerifiedParty)
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/identity/VerifiedParty.kt]: Similarly here

File: core/src/main/kotlin/net/corda/core/identity/VerifiedParty.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+data class VerifiedParty(val party: Party,
+                         val certificate: X509CertificateHolder,
+                         val certPath: CertPath) {
+    constructor(name: X500Name, owningKey: PublicKey, certificate: X509CertificateHolder, certPath: CertPath) : this(Party(name, owningKey), certificate, certPath)
+
+    init {
+        require(certPath.certificates.isNotEmpty()) { "Certificate path cannot be empty" }
+        val targetCert = certPath.certificates.first() as X509Certificate
+        val subjectX500Name = X500Name(targetCert.subjectDN.name)
+        require(subjectX500Name == certificate.subject) { "Certificate path must end with the subject ${certificate.subject}" }
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/identity/VerifiedParty.kt]: We can also check the `party.name`?

File: core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt [hunk 4]
```diff
@@ -22,12 +22,12 @@ interface IdentityService {
...
+     * @param verifiedParty a party representing a legal entity.
      * @throws IllegalArgumentException if the certificate path is invalid, or if there is already an existing
      * certificate chain for the anonymous party.
      */
     @Throws(CertificateExpiredException::class, CertificateNotYetValidException::class, InvalidAlgorithmParameterException::class)
-    fun registerIdentity(party: PartyAndCertificate)
+    fun registerIdentity(verifiedParty: VerifiedParty)
 
     /**
      * Verify and then store an anonymous identity.
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt]: This needs `@return`

File: core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt [hunk 5]
```diff
@@ -39,7 +39,7 @@ interface IdentityService {
      */
     @Throws(CertificateExpiredException::class, CertificateNotYetValidException::class, InvalidAlgorithmParameterException::class)
     @Deprecated("Use verifyAndRegisterAnonymousIdentity() instead, which is the same function with a better name")
-    fun registerAnonymousIdentity(anonymousIdentity: AnonymousPartyAndPath, party: Party): PartyAndCertificate
+    fun registerAnonymousIdentity(anonymousIdentity: VerifiedAnonymousParty, party: Party): VerifiedParty
 
     /**
      * Verify and then store an anonymous identity.
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt]: ditto

File: core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt [hunk 6]
```diff
@@ -50,7 +50,7 @@ interface IdentityService {
      * certificate chain for the anonymous party.
      */
     @Throws(CertificateExpiredException::class, CertificateNotYetValidException::class, InvalidAlgorithmParameterException::class)
-    fun verifyAndRegisterAnonymousIdentity(anonymousIdentity: AnonymousPartyAndPath, wellKnownIdentity: Party): PartyAndCertificate
+    fun verifyAndRegisterAnonymousIdentity(anonymousIdentity: VerifiedAnonymousParty, wellKnownIdentity: Party): VerifiedParty
 
     /**
      * Verify an anonymous identity.
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt]: Perhaps elaborate to say that this converts an anonymous verified party + a unverified party -> verified party.

File: node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt [hunk 7]
```diff
@@ -59,31 +59,31 @@ class InMemoryIdentityService(identities: Iterable<PartyAndCertificate> = emptyS
 
     // TODO: Check the certificate validation logic
     @Throws(CertificateExpiredException::class, CertificateNotYetValidException::class, InvalidAlgorithmParameterException::class)
-    override fun registerIdentity(party: PartyAndCertificate) {
-        require(party.certPath.certificates.isNotEmpty()) { "Certificate path must contain at least one certificate" }
+    override fun registerIdentity(verifiedParty: VerifiedParty) {
+        require(verifiedParty.certPath.certificates.isNotEmpty()) { "Certificate path must contain at least one certificate" }
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/identity/InMemoryIdentityService.kt]: Some of these checks are now redundant

File: core/src/main/kotlin/net/corda/core/identity/VerifiedParty.kt (hunk 1 above)
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/identity/VerifiedParty.kt]: The code that was going to use that function was actually removed before its PR was merged, removing it.

File: docs/source/changelog.rst [hunk 8]
```diff
@@ -57,6 +57,15 @@ UNRELEASED
...
      use to exclude core Corda JARs from being built into Cordapp fat JARs.
 
+* Changes to identity as part of confidential identities:
+
+  * ``PartyAndCertificate`` has been renamed to ``VerifiedParty``, and a corresponding ``AnonymousVerifiedParty`` class
+    added.
+
+  * ``registerAnonymousIdentity`` is now ``verifyAndRegisterAnonymousIdentity`` to indicate it performs sanity checks on
+    the identity provided. A separate ``verifyAnonymousIdentity`` function has been added for use-cases where verification
+    is performed indepdently of registration.
```
Comment (shamsasari) [inline docs/source/changelog.rst]: typo: indepdently

File: core/src/main/kotlin/net/corda/core/identity/VerifiedAnonymousParty.kt [hunk 9]
```diff
@@ -12,18 +12,18 @@ import java.security.cert.X509Certificate
...
 @CordaSerializable
-data class AnonymousPartyAndPath(
+data class VerifiedAnonymousParty(
         val party: AnonymousParty,
         val certPath: CertPath) {
     constructor(party: PublicKey, certPath: CertPath)
             : this(AnonymousParty(party), certPath)
     init {
-        require(certPath.certificates.isNotEmpty()) { "Certificate path cannot be empty" }
+        require(certPath.certificates.isNotEmpty()) { "Certificate path must contain at least one certificate" }
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/identity/VerifiedAnonymousParty.kt]: Is it possible to check the party's public key matches the relevant one from `certPath`?

**Comment by rnicoll** [comment]
Going to close this for now and come back to it once #1073 is in; in particular it may be that the classes need to be `VerifiableParty` and `VerifiableAnonymousParty`, as inherently the init{} checks can be bypassed by Kryo, which makes these names imply checks which aren't true.

**Comment by shamsasari** [comment]
We can write our own Kryo serialisers which invokes the public c'tors, so that's not really an issue.

**Pull request closed without merging**